From 88b2ce5b59ad834fc415b426beb6e77072cf28cc Mon Sep 17 00:00:00 2001
From: Chris Feist <cfeist@redhat.com>
Date: Thu, 12 Jun 2014 19:57:34 -0500
Subject: [PATCH 4/5] Only propagate/activate cluster on last node when using
 --activate

- Make it clear in documentation that --activate should only be
  used with --sync
---
 ricci/ccs/ccs   | 31 +++++++++++++++++++++----------
 ricci/ccs/ccs.8 |  4 ++--
 2 files changed, 23 insertions(+), 12 deletions(-)

diff --git a/ricci/ccs/ccs b/ricci/ccs/ccs
index 5e7c657..ce55bd2 100755
--- a/ricci/ccs/ccs
+++ b/ricci/ccs/ccs
@@ -347,10 +347,10 @@ Cluster configuration system.
                         cluster.conf file
                         same cluster.conf
       --getschema       Print current cluster schema file (if using -h use
-      			schema from network, if using -f use local schema)
-      --sync            Sync config file to all nodes
-      --activate        Activate config on node (use this option with --sync
-                        to activate config on all nodes)
+                        schema from network, if using -f use local schema)
+      --sync [--activate]
+                        Sync config file to all nodes and optionally activating
+                        that configuration on all nodes
   -d, --debug           Display debugging information to help troubleshoot
                         connection issues with ricci
       --exp tag [location] [options]
@@ -668,15 +668,26 @@ def get_cluster_status():
     print xml
 
 def sync_cluster_conf():
-    global hostname
+    global hostname, activate
+    temp_activate = activate
+    activate = False
     xml = get_cluster_conf_xml()
     dom = minidom.parseString(xml).getElementsByTagName('cluster')[0]
     xml = dom.toprettyxml(indent='  ',newl='')
     dom = minidom.parseString(xml)
-    for node in dom.getElementsByTagName('clusternode'):
-        nodename = node.getAttribute("name")
-        hostname = nodename
-        set_cluster_conf(xml)
+    nodes = dom.getElementsByTagName('clusternode')
+    if len(nodes) > 0:
+	for node in nodes[0:-1]:
+	    nodename = node.getAttribute("name")
+	    hostname = nodename
+	    set_cluster_conf(xml)
+
+	# If we're using --activate, only propogate on the last node
+	activate = temp_activate
+	node = nodes[-1]
+	nodename = node.getAttribute("name")
+	hostname = nodename
+	set_cluster_conf(xml)
 
 def list_nodes():
     xml = get_cluster_conf_xml()
@@ -2218,7 +2229,7 @@ def set_cluster_conf(xml, increment = True):
             log_msg (send_ricci_command("cluster", "set_cluster.conf", hostname, ("propagate", "boolean", "true"),("cluster.conf","xml","",xml)))
         else:
             log_msg (send_ricci_command("cluster", "set_cluster.conf", hostname, ("cluster.conf","xml","",xml)))
-
+    
 # Returns 0 if verification succeeded, an error code if it didn't
 def verify_cluster_conf(xml):
     if (not os.path.isfile(CLUSTERRNG)):
diff --git a/ricci/ccs/ccs.8 b/ricci/ccs/ccs.8
index 4f2c459..65bd6d0 100644
--- a/ricci/ccs/ccs.8
+++ b/ricci/ccs/ccs.8
@@ -37,8 +37,8 @@ host's cluster.conf file have the identical
 .IP "--getschema"
 Print current cluster schema file (if using -h use
 schema from network, if using -f use local schema)
-.IP "--sync"
-Sync config file to all nodes
+.IP "--sync" ["--activate"]
+Sync config file to all nodes and optionally activating that configuration on all nodes
 .IP "--activate"
 Activate config on node (use this option with --sync
 to activate config on all nodes)
-- 
1.9.3

